src/zarr/storage/_cache.py

```
@@ -0,0 +1,532 @@
+        self.invalidate_values()
+
+    async def exists(self, key: str) -> bool:
+        """
+        Check if a key exists in the store.
+
+        This method first checks the cache for the key to avoid
+        unnecessary calls to the underlying store.
+        """
+        cache_key = key
+
+        # Check cache first
+        with self._mutex:
+            if cache_key in self._values_cache:
+                # Key exists in cache, so it exists in store
+                # Move to end to mark as recently used
+                self._values_cache.move_to_end(cache_key)
+                return True
+
+        # Not in cache, delegate to underlying store
+        return await self._store.exists(key)
+
+    async def _set(
+        self,
+        key: str,
+        value: Buffer,
+        exclusive: bool = False,
+        byte_range: tuple[int, int] | None = None,
+    ) -> None:
+        # Check if store is writable
+        self._check_writable()
+
+        await self._store.set(key, value)
+
+        # Update cache
+        self.invalidate_keys()
+        self.invalidate_values()
+        cache_key = key
+        self._cache_value(cache_key, value)
+
+    async def get(
+        self,
+        key: str,
+        prototype: BufferPrototype | None = None,
+        byte_range: ByteRequest | None = None,
+    ) -> Buffer | None:
+        # Use the cache for get operations
+        cache_key = key
+
+        # For byte_range requests, don't use cache for now (could be optimized later)
+        if byte_range is not None:
+            try:
+                if prototype is None:
+                    prototype = default_buffer_prototype()
+                return await self._store.get(key, prototype, byte_range)
+            except TypeError:
+                # Fallback to sync get from mapping - get full value and slice later
+                # For now, just return None for byte range requests on sync stores
+                return None
+
+        try:
```

A bit safter to do `if cache_key in self._values_cache` here instead of try/except. That way any other KeyErrors raised in the block will be correctly raised, instead of being silently caught.